### PR TITLE
fix: prevent data leakage in feature transformation (TASKS.md 3.8)

### PR DESCRIPTION
**Pull request by ipeirotis**
Three bugs fixed in Table2Vector and training pipelines:

1. Index misalignment: one-hot encoded DataFrames now explicitly receive
   the source DataFrame's index, preventing silent NaN introduction from
   pd.concat with mismatched indices. Also fixed in tabularize_vector()
   and evaluate/outliers.py.

2. MinMaxScaler leakage: added fit()/transform() API so scalers are
   fitted on training data only, not the full dataset before split.

3. OneHotEncoder leakage: same fix — encoder categories come from the
   training split only. Unseen test-set categories get all-zeros.

All training pipelines (Trainer, run_training_pipeline, worker.py,
train/task.py) now split before vectorizing: clean → split → fit on
train → transform both. The legacy vectorize_table() API is preserved
for scoring/evaluation where no train/test split is needed.

15 new tests in tests/features/test_data_leakage.py.

https://claude.ai/code/session_01NPsZmrrK8SavktGLxFzPit

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -124,23 +148,70 @@ def prepare_for_model(project_data, variable_types=None):
...
+        project_data, test_size=test_size, random_state=None
+    )
+
+    # 5-6. Fit on training data, transform both
+    vectorizer.fit(train_df)
+    X_train = vectorizer.transform(train_df).astype("float32")
+    X_test = vectorizer.transform(test_df).astype("float32")
+
+    # 7. Cardinalities (from the full cleaned data so all categories are counted)
+    cardinalities = [project_data[c].nunique() for c in project_data.columns]
```
[the remark this answers is not in the document]
Comment (ipeirotis) [reply main.py]: Fixed in f4ceb96. Added `Table2Vector.get_cardinalities()` that derives cardinalities from the fitted encoder's `categories_` (or 1 for numeric columns). All callers (`prepare_for_model`, `prepare_for_training`, `worker.py`, `train/task.py`) now use this instead of `nunique()` on the full dataset. 4 new tests added to verify encoder-only categories are counted.

File: main.py [hunk 2]
```diff
@@ -100,47 +131,89 @@ def prepare_for_model(project_data, variable_types=None):
...
+
+    # 4. Split *before* vectorization
+    train_df, test_df = train_test_split(
+        project_data, test_size=test_size, random_state=None
+    )
+
+    # 5-6. Fit on training data, transform both
+    vectorizer.fit(train_df)
+    X_train = vectorizer.transform(train_df).astype("float32")
+    X_test = vectorizer.transform(test_df).astype("float32")
```
[the remark this answers is not in the document]
Comment (ipeirotis) [reply main.py:241]: Fixed in b0decca. The fitted vectorizer is now persisted alongside the model via `joblib`:

- `save_model()` accepts an optional `vectorizer` and writes `vectorizer.joblib` next to the Keras checkpoint
- `load_vectorizer()` loads it back (returns `None` for old checkpoints — backward compatible)
- `evaluate` and `find_outliers` now load the saved vectorizer and use `transform()` instead of refitting via `prepare_for_model()`, guaranteeing the same column layout as training

3 new tests for the save/load roundtrip, backward compat (no vectorizer file), and verifying that scoring data with unseen categories produces the training-width encoding.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: main.py [hunk 3]
```diff
@@ -100,50 +132,92 @@ def prepare_for_model(project_data, variable_types=None):
...
+    )
+
+    # 4. Split *before* vectorization
+    train_df, test_df = train_test_split(
+        project_data, test_size=test_size, random_state=None
+    )
+
+    # 5-6. Fit on training data, transform both
+    vectorizer.fit(train_df)
+    X_train = vectorizer.transform(train_df).astype("float32")
```
[the remark this answers is not in the document]
Comment (ipeirotis) [reply main.py:240]: Fixed in 2e2721a. The `generate` command now loads the persisted vectorizer (same pattern as `evaluate`/`find_outliers`), falling back to `prepare_for_model()` for old checkpoints without a saved vectorizer.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: worker.py [hunk 4]
```diff
@@ -1079,15 +1079,23 @@ def _migrate_legacy(txn):
...
+        vectorizer.fit(train_df)
+        X_train = vectorizer.transform(train_df).astype('float32')
+        X_test = vectorizer.transform(test_df).astype('float32')
+
+        # We also need the full vectorized data for scoring later
+        vectorized_df = vectorizer.transform(process_df).astype('float32')
 
         # 7. Build and train autoencoder
-        cardinalities = [process_df[col].nunique() for col in process_df.columns]
+        cardinalities = vectorizer.get_cardinalities(process_df.columns)
```
[the remark this answers is not in the document]
Comment (ipeirotis) [reply worker.py:1096]: Fixed in 4e51d72. `get_cardinalities()` now enforces `max(n_categories, 2)` for categorical columns, preventing `log(1)=0` division-by-zero in loss normalization when a rare category lands entirely in the test split. New test added for the edge case.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: features/transform.py [hunk 5]
```diff
@@ -46,50 +46,140 @@ def __init__(self, variable_types):
...
+        Returns:
+            List[int] of per-column cardinalities.
+        """
+        cardinalities = []
+        for col in columns:
+            if col in self.one_hot_encoders:
+                n_cats = len(self.one_hot_encoders[col].categories_[0])
+                # Enforce minimum of 2 to prevent log(1)=0 division in
+                # loss normalization (TASKS.md 3.5).
+                cardinalities.append(max(n_cats, 2))
```
[the remark this answers is not in the document]
Comment (ipeirotis) [reply features/transform.py]: Good catch — reverted in 3e2c622. `get_cardinalities()` now reports the true encoder category count. The `log(1)=0` division-by-zero is instead guarded at the three normalization sites (`model/autoencoder.py`, `model/variational_autoencoder.py`, `evaluate/outliers.py`) with `max(np.log(c), 1e-10)`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: main.py [hunk 6]
```diff
@@ -395,9 +476,16 @@ def evaluate(
     variable_types = metadata.get("variable_types", {})
 
     logger.info(f"Transforming the data....")
-    project_data, vectorized_df, vectorizer, _ = prepare_for_model(
-        project_data, variable_types
-    )
+    saved_vectorizer = load_vectorizer(model_path)
+    if saved_vectorizer is not None:
+        # Use the training-fitted vectorizer so one-hot width matches the model
+        project_data = prepare_for_categorical(project_data)
```
[the remark this answers is not in the document]
Comment (ipeirotis) [reply main.py]: Fixed in 3e2c622. Added `_clean_for_saved_vectorizer()` that does `fillna("missing")` + `astype(str)` and selects exactly the columns the vectorizer was trained on — no Rule-of-9 re-filtering. All three inference paths (`evaluate`, `find_outliers`, `generate`) now use this instead of `prepare_for_categorical()`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: main.py [hunk 7]
```diff
@@ -81,9 +83,66 @@ def prepare_for_categorical(project_data):
...
+    project_data = project_data.fillna("missing")
+    project_data = project_data.astype(str)
+
+    # Keep exactly the columns the vectorizer knows about
+    trained_cols = (
+        list(vectorizer.one_hot_encoders.keys())
+        + list(vectorizer.min_max_scalers.keys())
+    )
+    available = [c for c in trained_cols if c in project_data.columns]
+    return project_data[available]
```
[the remark this answers is not in the document]
Comment (ipeirotis) [reply main.py]: Fixed in 61e632b. `_clean_for_saved_vectorizer()` now adds any missing trained columns filled with `"missing"` instead of silently dropping them. The vectorizer's `handle_unknown='ignore'` encodes these as all-zeros, preserving the full schema width the model expects.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: utils.py [hunk 8]
```diff
@@ -17,18 +18,44 @@ def set_seed(seed):
...
-def save_model(model, output_path):
+def save_model(model, output_path, vectorizer=None):
     if not os.path.exists(output_path):
         os.makedirs(output_path)
 
     filename = output_path + "autoencoder"
     model.save(filename, save_format="tf")
 
+    if vectorizer is not None:
+        joblib.dump(vectorizer, os.path.join(output_path, "vectorizer.joblib"))
```
[the remark this answers is not in the document]
Comment (ipeirotis) [reply utils.py:33]: Fixed in a073d4c. `save_model()` now normalizes `output_path` to end with `os.sep` and uses `os.path.join` for both the model and vectorizer paths. `load_vectorizer()` simplified to match.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: main.py [hunk 9]
```diff
@@ -81,9 +83,70 @@ def prepare_for_categorical(project_data):
...
+        cleaned_df with exactly the columns the vectorizer expects,
+        in the same order as training.
+    """
+    project_data = project_data.fillna("missing")
+    project_data = project_data.astype(str)
+
+    # Ensure exactly the columns the vectorizer was trained on are present
+    trained_cols = (
+        list(vectorizer.one_hot_encoders.keys())
+        + list(vectorizer.min_max_scalers.keys())
```
[the remark this answers is not in the document]
Comment (ipeirotis) [reply main.py]: Fixed in 3419b73. `_clean_for_saved_vectorizer()` now derives the trained column list from `vectorizer.var_types` (covering all column types, not just those with encoders/scalers), with a defensive fallback for any columns only tracked in the encoder/scaler dicts.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: main.py [hunk 10]
```diff
@@ -81,9 +83,70 @@ def prepare_for_categorical(project_data):
...
+    )
+    for col in trained_cols:
+        if col not in project_data.columns:
+            project_data[col] = "missing"
```
[the remark this answers is not in the document]
Comment (ipeirotis) [reply main.py]: Fixed in 3419b73. Missing numeric columns are now filled with `NaN` (not `"missing"` string), so `MinMaxScaler.transform()` handles them correctly via its existing non-NA row masking. Missing categorical columns still get `"missing"` as before.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: features/transform.py [hunk 11]
```diff
@@ -46,50 +46,135 @@ def __init__(self, variable_types):
...
+                scaler min/max values.
+
+        Returns:
+            self (for chaining).
+        """
+        for column in training_df.columns:
+            if column in self.var_types["numeric"]:
+                scaler = MinMaxScaler()
+                non_na = training_df[column].notna()
+                scaler.fit(training_df.loc[non_na, [column]])
```
[the remark this answers is not in the document]
Comment (ipeirotis) [reply features/transform.py]: Fixed in 8ce82bb. `fit()` now skips the scaler when all training values are NaN (column passes through unscaled). `_apply_transforms()` also guarded to check the scaler exists and has non-NA rows before transforming.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: features/transform.py [hunk 12]
```diff
@@ -46,50 +46,138 @@ def __init__(self, variable_types):
...
+                scaler = MinMaxScaler()
+                non_na = training_df[column].notna()
+                if non_na.sum() > 0:
+                    scaler.fit(training_df.loc[non_na, [column]])
+                    self.min_max_scalers[column] = scaler
+                # If all values are NaN in training, skip — column passes through unscaled
+            elif column in self.var_types["categorical"]:
+                encoder = OneHotEncoder(
+                    sparse_output=False, handle_unknown="ignore"
+                )
```
[the remark this answers is not in the document]
Comment (ipeirotis) [reply features/transform.py:74]: Good observation, but I believe this is better addressed as a separate follow-up rather than in this PR:

1. `handle_unknown='ignore'` (all-zeros for unseen categories) is pre-existing behavior — this PR didn't change how unknowns are encoded, only when the encoder is fitted
2. Adding an explicit "unknown" one-hot slot would change decoder output width, loss computation, and require model retraining — architecturally significant
3. With the saved vectorizer, the typical workflow (train and score on same dataset) ensures consistent categories, so all-zeros only occurs for genuinely novel categories in new data

This would be a good addition to TASKS.md as a model quality improvement (related to section 8 on missing data handling).

---
_Generated by [Claude Code](https://claude.ai/code)_
